ASAN_TRAP in WebCore::WorkletGlobalScope::WorkletGlobalScope

pratiksha-99 opened the pull request:
#### f834e56e78df4c0a34dd1e210203aaafc4746b4c
<pre>
ASAN_TRAP in WebCore::WorkletGlobalScope::WorkletGlobalScope
<a href="https://bugs.webkit.org/show_bug.cgi?id=284969">https://bugs.webkit.org/show_bug.cgi?id=284969</a>
<a href="https://rdar.apple.com/141683469">rdar://141683469</a>

Reviewed by Chris Dumez and Ryosuke Niwa.

Added a check for document having browsing context.

* LayoutTests/fast/worklets/worklet-global-scope-constructor-expected.txt: Added.
* LayoutTests/fast/worklets/worklet-global-scope-constructor.html: Added.
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/css/DOMCSSPaintWorklet.cpp:
(WebCore::PaintWorklet::addModule):

Canonical link: <a href="https://commits.webkit.org/288257@main">https://commits.webkit.org/288257@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf72028816fda2ae2fd1e565b5b5f0cc2aa87d06

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82261 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1925 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36327 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87393 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33322 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84366 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1996 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9807 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64113 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21863 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85330 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1391 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74852 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44391 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1293 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29033 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32363 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72592 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29657 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88749 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9567 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6830 "Found 1 new test failure: imported/w3c/web-platform-tests/media-source/mediasource-changetype.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72504 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9793 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70666 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71722 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15860 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14874 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/922 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12768 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9520 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15041 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9394 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12860 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11163 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/fast/worklets/worklet-global-scope-constructor.html:
```diff
@@ -0,0 +1,83 @@
+<!DOCTYPE html>
```
I believe you can simply the test to:
```html
<!DOCTYPE html>
<html lang="en">
<p>This test passes if WebKit does not crash.</p>
<body>
<script>
globalThis.testRunner?.dumpAsText();

const iframe = document.createElement('iframe');
document.body.appendChild(iframe);
const contentCSS = iframe.contentWindow.CSS;
iframe.remove();
const moduleURL = {};
const options = {};
contentCSS.paintWorklet.addModule(moduleURL, options);

</script>
</body>
</html>
```

cdumez commented on Source/WebCore/worklets/PaintWorkletGlobalScope.cpp:
```diff
@@ -42,7 +42,7 @@ WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(PaintWorkletGlobalScope);
 {
     RefPtr<VM> vm = VM::tryCreate();
-    if (!vm)
+    if (!vm || !document.sessionID())
```
Checking for `document.sessionID()` is likely not the right thing to do here. How about `!document.frame()` instead?

rniwa replied:
I don't think we really care either way. We're not shipping PaintWorklet anytime soon, and the spec doesn't mention browsing context at all so it won't be spec compliant anyway. It's likely a spec bug.

cdumez replied:
I care. We check that documents have frames all the time in Web APIs. Checking that a document has a session ID? no. The only reason the document doesn't have a sessionID is because it doesn't have a frame likely.

rniwa replied:
> I care. We check that documents have frames all the time in Web APIs. Checking that a document has a session ID? no. The only reason the document doesn't have a sessionID is because it doesn't have a frame likely.

But there is no basis in the spec to do that.

pratiksha-99 replied:
I am unable to verify due to WTR crashes on recent ASAN builds. If you are sure, I can update the PR.

rniwa replied:
@pratiksha-99 : For expediency, you should just go ahead & do that. But looking at the code more closely, it's probably not the right place to add a check because returning `nullptr` here results in `OutOfMemoryError` to be emitted in `PaintWorklet::addModule` in `DOMCSSPaintWorklet.cpp`. That's certainly not the right type of error to throw. We should add a check for `frame()` returning null in `PaintWorklet::addModule` like this (we should use `RefPtr` in place of `auto*` for document):
```cpp
void PaintWorklet::addModule(const String& moduleURL, WorkletOptions&&, DOMPromiseDeferred<void>&& promise)
{
    RefPtr document = this->document();
    if (!document) {
        promise.reject(Exception { ExceptionCode::InvalidStateError, "This frame is detached"_s });
        return;
    }
    if (!document->hasBrowsingContext()) {
        promise.reject(Exception { ExceptionCode::InvalidStateError, "This document does not have a browsing context"_s });
        return;
    }
```

cdumez reviewed (changes requested):
Please check the frame.

rniwa commented on Source/WebCore/worklets/PaintWorkletGlobalScope.cpp:
```diff
@@ -42,6 +42,7 @@ WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(PaintWorkletGlobalScope);
 RefPtr<PaintWorkletGlobalScope> PaintWorkletGlobalScope::tryCreate(Document& document, ScriptSourceCode&& code)
 {
     RefPtr<VM> vm = VM::tryCreate();
+    JSC::JSLockHolder lock(*vm);
```
This needs to be after the nullptr check for vm below:

pratiksha-99 committed:
ASAN_TRAP in WebCore::WorkletGlobalScope::WorkletGlobalScope
https://bugs.webkit.org/show_bug.cgi?id=284969
rdar://141683469

Reviewed by Chris Dumez and Ryosuke Niwa.

Added a check for document having browsing context.

* LayoutTests/fast/worklets/worklet-global-scope-constructor-expected.txt: Added.
* LayoutTests/fast/worklets/worklet-global-scope-constructor.html: Added.
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/css/DOMCSSPaintWorklet.cpp:
(WebCore::PaintWorklet::addModule):

Canonical link: https://commits.webkit.org/288257@main
